src/form-data-wrapper.js

```diff
@@ -14,22 +14,148 @@
  * limitations under the License.
  */
 
-import {dev} from './log';
 import {getFormAsObject} from './form';
 import {map} from './utils/object';
 
+/**
+ * Create a form data wrapper. The wrapper is necessary to provide a common
+ * API for FormData objects on all browsers. For example, not all browsers
+ * support the FormData#entries or FormData#delete functions.
+ *
+ * @param {!HTMLFormElement=} opt_form
+ * @return {!FormDataWrapperInterface}
+ */
+export function createFormDataWrapper(opt_form) {
+  if (FormData.prototype.entries && FormData.prototype.delete) {
+    return new NativeFormDataWrapper(opt_form);
+  } else {
+    return new PolyfillFormDataWrapper(opt_form);
+  }
+}
+
+/**
+ * Check if the given object is a FormDataWrapper instance
+ * @param {*} o
+ * @return {boolean} True if the object is a FormDataWrapper instance.
+ */
+export function isFormDataWrapper(o) {
+  // instanceof doesn't work as expected, so we detect with duck-typing.
+  return !!o && typeof o.getFormData == 'function';
+}
+
+/**
+ * A polyfill wrapper for a `FormData` object.
+ *
+ * If there's no native `FormData#entries`, chances are there are no native
+ * methods to read the content of the `FormData` after construction, so the
+ * only way to implement `entries` in this class is to capture the fields in
+ * the form passed to the constructor (and the arguments passed to the
+ * `append` method).
+ *
+ * For more details on this, see http://mdn.io/FormData.
+ *
+ * @implements {FormDataWrapperInterface}
+ * @visibleForTesting
+ */
+export class PolyfillFormDataWrapper {
+  /** @override */
+  constructor(opt_form = undefined) {
+    /** @private @const {!Object<string, !Array<string>>} */
+    this.fieldValues_ = opt_form ? getFormAsObject(opt_form) : map();
+  }
+
+  /** @override */
+  append(name, value) {
+    // Coercion to string is required to match
+    // the native FormData.append behavior
+    const nameString = String(name);
+    this.fieldValues_[nameString] = this.fieldValues_[nameString] || [];
+    this.fieldValues_[nameString].push(String(value));
+  }
+
+  /** @override */
+  delete(name) {
+    delete this.fieldValues_[name];
+  }
+
+  /** @override */
+  entries() {
+    const fieldEntries = [];
+    Object.keys(this.fieldValues_).forEach(name => {
+      const values = this.fieldValues_[name];
+      values.forEach(value => fieldEntries.push([name, value]));
+    });
+
+    // Generator functions are not supported by the current Babel configuration,
```

Comment:
Fantastic!

Are we certain this is not generating client code?

Reply:
This shouldn't generate any client code because this function manually implements a generator instead of using `*function` generator syntax.

Reply:
Perfect.